### PR TITLE
Add proper cache behavior for lift.jit

### PR DESCRIPTION
**Pull request by jheek**
Fixes cache misses in lift.jit

NOTE: how to cache jit in linen transforms still needs additional changes

### CODE REVIEW

File: flax/core/lift.py [hunk 1]
```diff
@@ -715,21 +723,39 @@ def jit(fn: Callable[..., Any],
...
+  # where scope_fn or repack_fn actually produce non-identical results.
+  scope_fn = None  # type: Callable
+  repack_fn = None  # type: Callable
+  @functools.partial(jax.jit,
+                     static_argnums=static_argnums,
+                     donate_argnums=donate_argnums,
+                     device=device, backend=backend)
+  @functools.wraps(fn)
+  def jitted(fingerprint, variable_groups, rng_groups, *args):
+    # fingerprint is only used to differentiate the cache signature for cases
```
Comment (levskaya) [inline flax/core/lift.py:742]: would it be useful to mark scope_fn, repack_fn 'nonlocal' here as well just to emphasize what we're doing for future readers?

**Review by levskaya — APPROVED** [review]
A clever hack!  I guess we could do this to make a useful version of my old pmap hack as well.

File: flax/core/lift.py (hunk 1 above)
Comment (jheek) [reply flax/core/lift.py:742]: > A clever hack! what exactly does your note in the comment above mean? is this unsound when nested inside other transforms?

Just like jit the cache doesn't trigger if you call lift.jit multiple times which is what transforms.py ends up doing.

**Comment by jheek** [comment]
> A clever hack! what exactly does your note in the comment above mean? is this unsound when nested inside other transforms?

Just like jit the cache doesn't trigger if you call lift.jit multiple times which is what transforms.py ends up doing.

**Pull request merged**